lib/phoenix_component/macro_component.ex

```diff
@@ -0,0 +1,321 @@
+
+  > #### Limitations {: .warning}
+  > The AST is not whitespace preserving, so in cases where you return a modified AST,
+  > the original whitespace between attributes is lost.
+  >
+  > Also, macro components can currently only contain simple HTML. Any interpolation
+  > like `<%= @foo %>` or components inside are not supported.
+
+  ## Example: a compile-time markdown renderer
+
+  Let's say we want to create a macro component that renders markdown as HTML at
+  compile time. First, we need some library that actually converts the markdown to
+  HTML. For this example, we use [`mdex`](https://hex.pm/packages/mdex).
+
+  We start by defining the module for the macro component:
+
+  ```elixir
+  defmodule MyAppWeb.MarkdownComponent do
+    @behaviour Phoenix.Component.MacroComponent
+
+    @impl true
+    def transform({"pre", attrs, children}, _meta) do
+      markdown = Phoenix.Component.MacroComponent.to_string(children)
+      html_doc = MDEx.to_html!(markdown)
+
+      {"div", attrs, [html_doc]}
+    end
+  end
+  ```
+
+  That's it. Since the div could contain nested elements, for example when using
+  an HTML code block, we need to convert the children to a string first, using the
+  `Phoenix.Component.MacroComponent.to_string/1` function.
+
+  Then, we can simply replace the element's contents with the returned HTML string from
+  MDEx.
+
+  We can now use the macro component inside our HEEx templates:
+
+      defmodule MyAppWeb.ExampleLiveView do
+        use MyAppWeb, :live_view
+
+        def render(assigns) do
+          ~H\"\"\"
+          <pre :type={ColocatedDemoWeb.Markdown} class="prose mt-8">
+          ## Hello World
+
+          This is some markdown!
+
+          ```elixir
+          defmodule Hello do
+            def world do
+              IO.puts "Hello, world!"
+            end
+          end
+          ```
+          </pre>
+          \"\"\"
+        end
+      end
+
+  Note: this example uses the `prose` class from TailwindCSS for styling.
+
+  One trick to prevent issues with extra whitespace is that we use a `<pre>` tag in the LiveView
+  template, which prevents the `Phoenix.LiveView.HTMLFormatter` from indenting the contents, which
+  would mess with the markdown parsing. When rendering, we replace it with a `<div>` tag in the
+  macro component.
+  """
+
+  @type tag :: binary()
+  @type attributes :: %{atom() => term()}
+  @type children :: [heex_ast()]
+  @type tag_meta :: %{closing: :self | :void}
+  @type heex_ast :: {tag(), attributes(), children(), tag_meta()} | binary()
+  @type transform_meta :: %{env: Macro.Env.t()}
+
+  @callback transform(heex_ast :: heex_ast(), meta :: transform_meta()) ::
+              {:ok, heex_ast()} | {:ok, heex_ast(), data :: term()}
+
+  @doc """
+  Returns the stored data from macro components that returned `{:ok, ast, data}` in
+  the format `%{module => [data]}` where module is the macro component from `:type`.
+  """
+  def get_data(component_module) do
```

Comment:
Do we want to pass the macro component name here as well? So we return only the keys it is supposed to access?

Reply:
adjusted it to be `get_data(component_module, macro_component)`

https://github.com/phoenixframework/phoenix_live_view/pull/3810/commits/bbbfadbba9d74fddfc127ad266c4bd92136b5779